### EDA dataset: Amazon's Top 50 bestselling books from 2009 to 2019
#### TABLE OF CONTENT
#### 0 PROLOGUE
#### 1 IMPORTING LIBRARIES
#### 2 DATA DESCRIPTION AND DATA CLEANING
#### 2.1 Data types
#### 2.2 Missing values
#### 2.3 Duplicates
#### 2.4 Categorical data
#### 2.5 Numeric data

###  PROLOGUE

### In that work, exploratory data analysis has been carried out dataset Amazon's Top 50 bestselling books from 2009 to 2019.
##### FEATURES:
#### Name - Name of the Book
#### Author - The author of the Book
#### User Rating - Amazon User Rating
#### Reviews - Number of written reviews on amazon
#### Price - The price of the book (As at 13/10/2020)
#### Year - The Year(s) it ranked on the bestseller
#### Genre - Whether fiction or non-fiction

#### ANSWER TO THE FOLLOWING QUESTIONS ARE GIVEN:
#### Which author's books receive the highest average rating (top authors).
#### Which author has written the most bestsellers (top authors).
#### Which book has the most reviews (top books).
#### Which genres become bestsellers more often.
#### Are genres different by rating.

### IMPORTING LIBRARIES

#### LIBRARIES:
#### Library pandas will be required to work with data in tabular representation.
#### Library numpy will be required to round the data in the correlation matrix.
#### Library matplotlib, seaborn, plotly required for data visualization.

In [1]:
import pandas as pd
import numpy as np


####  DATA DESCRIPTION AND DATA CLEANING

In this block, exploratory data analysis will be carried out, data types, missing values, duplicates, measures of the central trend, and also the relationship between variables will be described.

### Data types

In [3]:
# Reading data
data = pd.read_csv('bestsellers with categories.csv') # Loading data
data.head() # Displaying the first 5 lines of the dataset

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

In [3]:
data.shape #shape of data

(550, 7)

In [6]:
df = pd.DataFrame(data) #convert data into dataframe

In [5]:
#to get type of data
df.dtypes

Name            object
Author          object
User Rating    float64
Reviews          int64
Price            int64
Year             int64
Genre           object
dtype: object

#### Data contains 3 categorical columns and 4 columns contains numeric values. Let's convert the genre to the category data type, leaving the year as a number.

In [6]:
# Change the data type
df.Genre = df.Genre.astype('category')

In [7]:
df.dtypes

Name             object
Author           object
User Rating     float64
Reviews           int64
Price             int64
Year              int64
Genre          category
dtype: object

### Missing values

 Let's calculate number of missing/null values in data

In [8]:
df.isnull().sum()

Name           0
Author         0
User Rating    0
Reviews        0
Price          0
Year           0
Genre          0
dtype: int64

CONCLUSION: The data has no missing values, so no further transformations are required.

### Duplicates

In [9]:
# Forming categorical columns
col_data = list(df.select_dtypes(exclude=('int64', 'float64')).columns)
print(f'Сolumns without numeric data: {", ".join(col_data)}.')

Сolumns without numeric data: Name, Author, Genre.


In [10]:
#lets see duplicates
#df.duplicated().any()
# Let's see the duplicates in the columns
for col in col_data:
    if df[col].duplicated().any() == True:
        print (f'Column {col} contains duplicates.')
    else:
        print (f'Column {col} does not contain duplicates.')

Column Name contains duplicates.
Column Author contains duplicates.
Column Genre contains duplicates.


In [11]:
df[df['Name'].duplicated()]['Name'].value_counts().head()

Publication Manual of the American Psychological Association, 6th Edition       9
StrengthsFinder 2.0                                                             8
Oh, the Places You'll Go!                                                       7
The Very Hungry Caterpillar                                                     6
The 7 Habits of Highly Effective People: Powerful Lessons in Personal Change    6
Name: Name, dtype: int64

In [12]:
#number of duplicatesStrengthsFinder 2.0
df[df['Name'].str.contains("StrengthsFinder 2.0", case = False)]['Name'].count()

9

In [13]:
# Check if spelling errors have been hit
for col in col_data:
    print(f'Before {col}: {len(set(df[col]))} After {col}: {len(set(df[col].str.title().str.strip()))}')

Before Name: 351 After Name: 350
Before Author: 248 After Author: 248
Before Genre: 2 After Genre: 2


In [14]:
#Indeed, there were typing errors in the Name column. Let's fix it.
# Correct the errors
df.Name = df.Name.str.title().str.strip()

In [15]:
# Check if the changes have passed
for col in col_data:
    print(f'Before {col}: {len(set(df[col]))} After {col}: {len(set(df[col].str.title().str.strip()))}')

Before Name: 350 After Name: 350
Before Author: 248 After Author: 248
Before Genre: 2 After Genre: 2


In [16]:
#Let's check if there are the same author names but with different spellings.
authors = df.Author.sort_values().unique()
authors

array(['Abraham Verghese', 'Adam Gasiewski', 'Adam Mansbach', 'Adir Levy',
       'Admiral William H. McRaven', 'Adult Coloring Book Designs',
       'Alan Moore', 'Alex Michaelides', 'Alice Schertle', 'Allie Brosh',
       'American Psychiatric Association',
       'American Psychological Association', 'Amor Towles', 'Amy Ramos',
       'Amy Shields', 'Andy Weir', 'Angie Grace', 'Angie Thomas',
       'Ann Voskamp', 'Ann Whitford Paul', 'Anthony Bourdain',
       'Anthony Doerr', 'Atul Gawande', 'Audrey Niffenegger',
       'B. J. Novak', 'Bessel van der Kolk M.D.', 'Bill Martin Jr.',
       "Bill O'Reilly", 'Bill Simmons', 'Blue Star Coloring',
       'Bob Woodward', 'Brandon Stanton', 'Brené Brown',
       'Brian Kilmeade', 'Bruce Springsteen', 'Carol S. Dweck',
       'Celeste Ng', 'Charlaine Harris', 'Charles Duhigg',
       'Charles Krauthammer', 'Cheryl Strayed', 'Chip Gaines',
       'Chip Heath', 'Chris Cleave', 'Chris Kyle', 'Chrissy Teigen',
       'Christina Baker Kline', 

In [17]:
# Replace the names of the authors with the correct ones
df = df.replace('George R. R. Martin', 'George R.R. Martin')
df = df.replace('J. K. Rowling', 'J.K. Rowling')

In [18]:
# Check if the changes have passed
for col in col_data:
    print(f'Before {col}: {len(set(df[col]))} After {col}: {len(set(df[col].str.title().str.strip()))}')

Before Name: 350 After Name: 350
Before Author: 246 After Author: 246
Before Genre: 2 After Genre: 2


In [19]:
# Check the column Genre
genre = df.Genre.unique()
genre

array(['Non Fiction', 'Fiction'], dtype=object)

In [20]:
#Everything is fine with genres.
# Check the column Year
years = list(df.Year.unique())
sorted(years)

[2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]

Everything is fine with years.
In addition, there may be duplicate rows in the data. Let's display the last 5 lines of the dataframe.

In [21]:
df.tail()

Name       Author  \
545       Wrecking Ball (Diary Of A Wimpy Kid Book 14)  Jeff Kinney   
546  You Are A Badass: How To Stop Doubting Your Gr...  Jen Sincero   
547  You Are A Badass: How To Stop Doubting Your Gr...  Jen Sincero   
548  You Are A Badass: How To Stop Doubting Your Gr...  Jen Sincero   
549  You Are A Badass: How To Stop Doubting Your Gr...  Jen Sincero   

     User Rating  Reviews  Price  Year        Genre  
545          4.9     9413      8  2019      Fiction  
546          4.7    14331      8  2016  Non Fiction  
547          4.7    14331      8  2017  Non Fiction  
548          4.7    14331      8  2018  Non Fiction  
549          4.7    14331      8  2019  Non Fiction

In addition, it is worth noting that there are duplicate rows in the data, but with different years. Let's remove duplicates. Let's leave the data without years for later analysis.

In [22]:
# Delete the year column
del df['Year']

In [23]:
# Remove duplicates and check how many books are left in the data
df = df.drop_duplicates(keep='first')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 361 entries, 0 to 546
Data columns (total 6 columns):
Name           361 non-null object
Author         361 non-null object
User Rating    361 non-null float64
Reviews        361 non-null int64
Price          361 non-null int64
Genre          361 non-null object
dtypes: float64(1), int64(2), object(3)
memory usage: 19.7+ KB


In [24]:
# Let's count the number of each book in the data
# Сheck with the example of a book The Help.
df[df['Name']== 'The Help']

Name            Author  User Rating  Reviews  Price    Genre
402  The Help  Kathryn Stockett          4.8    13871      6  Fiction
404  The Help  Kathryn Stockett          4.8    13871      8  Fiction
405  The Help  Kathryn Stockett          4.8    13871      7  Fiction

 Records are duplicated by price. Apparently, in different years the book costs differently, which is quite logical taking into account inflation, demand, etc. Let's leave only the last entries.

In [25]:
# Drop duplicates
df= df.drop_duplicates(subset='Name', keep='last')
df

Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules For Life: An Antidote To Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
538  Winter Of The World: Book Two Of The Century T...   
539  Women Food And God: An Unexpected Path To Almo...   
540                                             Wonder   
545       Wrecking Ball (Diary Of A Wimpy Kid Book 14)   
546  You Are A Badass: How To Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price        Genre  
0                    JJ Smith          4.7    17350      8  Non Fiction  
1                Stephen King          4.6     2052     22      Fiction  
2          Jordan B. Peterson          4.7    18979     15  Non Fiction  
3               George Orwell          4.7    21424      6      Fiction  
4    National Geographic Kids          4.8     7665     12  Non Fiction  
..                        ...          ...      ...    ...          ...  
538               Ken Follett          4.5    10760     15      Fiction  
539               Geneen Roth          4.2     1302     11  Non Fiction  
540             R. J. Palacio          4.8    21625      9      Fiction  
545               Jeff Kinney          4.9     9413      8      Fiction  
546               Jen Sincero          4.7    14331      8  Non Fiction  

[350 rows x 6 columns]

In [26]:
len(df['Author'].unique()) # number of Authors

246

CONCLUSION: Thus, the data contains 350 different books written by 246 authors. All books are presented in two categories (Non Fiction, Fiction).

### Categorical data
In this section, we will analyze categorical data and answer the following questions:
Which author's books receive the highest average rating (top authors).

Which author has written the most bestsellers (top authors).

Which genres become bestsellers more often.

Which book has the most reviews (top books).

Let's create new dataframes based on the average rating of authors, the number of books written by authors, the number of reviews for books and the number of books by genre.

In [27]:
df.groupby('Author')[['User Rating']].mean().sort_values('User Rating', ascending=False).head(13).reset_index()


Author  User Rating
0          Nathan W. Pyle          4.9
1          Patrick Thorpe          4.9
2              Eric Carle          4.9
3   Emily Winfield Martin          4.9
4             Chip Gaines          4.9
5              Jill Twiss          4.9
6           Rush Limbaugh          4.9
7    Sherri Duskey Rinker          4.9
8          Alice Schertle          4.9
9              Pete Souza          4.9
10            Sarah Young          4.9
11     Lin-Manuel Miranda          4.9
12        Bill Martin Jr.          4.9

In [34]:

df.groupby('Author')[['Name']].count().sort_values('Name', ascending=False).head(10).reset_index()

Author  Name
0       Jeff Kinney    12
1      Rick Riordan    10
2      J.K. Rowling     8
3   Stephenie Meyer     7
4        Dav Pilkey     6
5     Bill O'Reilly     6
6      John Grisham     5
7         E L James     5
8   Suzanne Collins     5
9  Charlaine Harris     4

In [35]:
df.groupby('Name')[['Reviews']].sum().sort_values('Reviews', ascending = False).reset_index()

Name  Reviews
0                              Where The Crawdads Sing    87841
1                                The Girl On The Train    79446
2                                             Becoming    61133
3                                            Gone Girl    57271
4                               The Fault In Our Stars    50482
..                                                 ...      ...
345  The Daily Show With Jon Stewart Presents Earth...      440
346                             True Compass: A Memoir      438
347                    George Washington'S Sacred Fire      408
348  Soul Healing Miracles: Ancient And New Sacred ...      220
349  Divine Soul Mind Body Healing And Transmission...       37

[350 rows x 2 columns]

In [30]:
df.groupby('Genre')[['Name']].count().sort_values('Name', ascending=False).reset_index()

Genre  Name
0  Non Fiction   190
1      Fiction   160

INSIGHT: By analyzing the categorical data, it is established:
The following 13 authors have the highest rating: Nathan W. Pyle, Patrick Thorpe, Eric Carle, Emily Winfield Martin, Chip Gaines, Jill Twiss, Rush Limbaugh, Sherri Duskey Rinker, Alice Schertle, Pete Souza, Sarah Young, Lin-Manuel Miranda, Bill Martin Jr., Dav Pilkey. The average rating for their works was 4.9. When buying a new book, you should pay attention to these authors.

Authors who have written more bestsellers: Jeff Kinney - 12 books, Rick Riordan - 10 books, J.K. Rowling - 8 books, Stephenie Meyer - 7 books, Dav Pilkey - 6 books, Bill O'Reilly - 6 books, John Grisham - 5 books, E L James - 5 books, Suzanne Collins - 5 books, Charlaine Harris - 4 books. These authors always have something to read.

Books with the most reviews: Where The Crawdads Sing - 87841 Reviews, The Girl On The Train - 79446 Reviews, Becoming - 61133 Reviews, Gone Girl - 57271 Reviews, The Fault In Our Stars - 50482 Reviews. It's definitely worth reading the book Where The Crawdads Sing, it's not for nothing that it is the most talked about.

Non-fiction is more likely to become a bestseller.

### Numeric data

Let's look at the measures of the central trend that characterize the distribution of numerical values in the data. We render a box-plot for each numeric variable.

In [31]:
# Displaying measures of the central trend
df.describe()

User Rating       Reviews       Price
count   350.000000    350.000000  350.000000
mean      4.608857   9804.605714   12.925714
std       0.226993  10885.017686   10.003161
min       3.300000     37.000000    0.000000
25%       4.500000   3435.250000    7.250000
50%       4.600000   6328.000000   11.000000
75%       4.800000  11510.250000   16.000000
max       4.900000  87841.000000  105.000000

INSIGHT: By analyzing the numeric data, it is established:
User Rating:
Average and median book ratings are 4.6.

Reviews:
The data has a wide range.

Price:
There are books that cost much higher than the average, as well as books with a cost of 0, which is strange. Either the book is given for free or this is an error.

In [32]:
df.corr()

User Rating   Reviews     Price
User Rating     1.000000 -0.055478 -0.028228
Reviews        -0.055478  1.000000 -0.045705
Price          -0.028228 -0.045705  1.000000

INSIGHT: Based on the constructed correlation matrix, it can be seen that the data does not contain any positive or negative linear relationship between the rating, reviews and the price of books.

###  FINAL CONCLUSIONS

In the course of the analysis, it was established which authors receive the highest ratings from readers, which authors have written the most bestsellers, which books receive the most reviews from readers. In addition, it was found that non-fiction literature is becoming more often a bestseller, but users also like fiction more.